libbeat/ebpf/watcher_linux.go

```diff
@@ -0,0 +1,183 @@
+type client struct {
+	name    string
+	mask    EventMask
+	records chan ebpfevents.Record
+}
+
+// EventMask is a mask of epbevents.EventType which is used to control which event types clients will receive.
+type EventMask uint64
+
+// Watcher observes kernel events, using ebpf probes from the ebpfevents library, and sends the
+// events to subscribing clients.
+//
+// A single global watcher can exist, and can deliver events to multiple clients. Clients subscribe
+// to the watcher, and all ebpf events that match their mask will be sent to their channel.
+type Watcher struct {
+	sync.Mutex
+	cancel  context.CancelFunc
+	loader  *ebpfevents.Loader
+	clients map[string]client
+	status  status
+}
+
+type status int
+
+const (
+	stopped status = iota
+	started
+)
+
+// GetWatcher creates the watcher, if required, and returns a reference to the global Watcher.
+func GetWatcher() (*Watcher, error) {
+	gWatcher.Lock()
+	defer gWatcher.Unlock()
+
+	// Try to load the probe once on startup so consumers can error out.
+	gWatcherOnce.Do(func() {
+		if gWatcher.status == stopped {
+			l, err := ebpfevents.NewLoader()
+			if err != nil {
+				gWatcherErr = fmt.Errorf("init ebpf loader: %w", err)
+				return
+			}
+			_ = l.Close()
+		}
+	})
+
+	return &gWatcher, gWatcherErr
+}
+
+// Subscribe to receive events from the watcher.
+func (w *Watcher) Subscribe(clientName string, events EventMask) <-chan ebpfevents.Record {
+	w.Lock()
+	defer w.Unlock()
+
+	if w.status == stopped {
+		startLocked()
+	}
+
+	w.clients[clientName] = client{
+		name:    clientName,
+		mask:    events,
+		records: make(chan ebpfevents.Record, w.loader.BufferLen()),
+	}
+
+	return w.clients[clientName].records
+}
+
+// Unsubscribe the client with the given name.
+func (w *Watcher) Unsubscribe(clientName string) {
+	w.Lock()
+	defer w.Unlock()
+
+	delete(w.clients, clientName)
+
+	if w.nclients() == 0 {
+		stopLocked()
+	}
+}
+
+func startLocked() {
+	if gWatcher.status == started {
+		return
+	}
+
+	loader, err := ebpfevents.NewLoader()
+	if err != nil {
+		gWatcherErr = fmt.Errorf("start ebpf loader: %w", err)
+		return
+	}
+
+	gWatcher.loader = loader
+	gWatcher.clients = make(map[string]client)
+
+	records := make(chan ebpfevents.Record, loader.BufferLen())
+	var ctx context.Context
+	ctx, gWatcher.cancel = context.WithCancel(context.Background())
+
+	go gWatcher.loader.EventLoop(ctx, records)
+	go func(ctx context.Context) {
+		for {
+			select {
+			case record := <-records:
+				if record.Error != nil {
+					for _, client := range gWatcher.clients {
+						client.records <- record
+					}
+					continue
+				}
+				for _, client := range gWatcher.clients {
+					if client.mask&EventMask(record.Event.Type) != 0 {
+						client.records <- record
+					}
+				}
+				continue
+			case <-ctx.Done():
+				return
+			}
+		}
+	}(ctx)
+
+	gWatcher.status = started
+}
+
+func stopLocked() {
+	if gWatcher.status == stopped {
+		return
+	}
+	gWatcher.close()
+	gWatcher.status = stopped
+}
```

Comment:
This feels very much like it should be a method on `Watcher`.

Reply:
I've added these to Watcher